File: shopping-cart/python/tests/product_builder.py
from __future__ import annotations

from typing import Optional

from shopping_cart import DiscountPolicy, Money, Product


class ProductBuilder:
    def __init__(self) -> None:
        self._sku = "SKU-DEFAULT"
        self._name = "Widget"
        self._unit_price = Money(10)
        self._discount_policy: Optional[DiscountPolicy] = None

    def with_sku(self, sku: str) -> "ProductBuilder":
        self._sku = sku
        return self

    def named(self, name: str) -> "ProductBuilder":
        self._name = name
        return self

    def priced_at(self, amount: float | str | int) -> "ProductBuilder":
        self._unit_price = Money(amount)
        return self

    def with_discount(self, policy: DiscountPolicy) -> "ProductBuilder":
        self._discount_policy = policy
        return self

    def build(self) -> Product:
        if self._discount_policy is None:
            return Product(self._sku, self._name, self._unit_price)
        return Product(self._sku, self._name, self._unit_price, self._discount_policy)

File: shopping-cart/typescript/tests/ProductBuilder.ts
import type { DiscountPolicy } from '../src/DiscountPolicy.js';
import { Money } from '../src/Money.js';
import { Product } from '../src/Product.js';

export class ProductBuilder {
  private _sku = 'SKU-DEFAULT';
  private _name = 'Widget';
  private _unitPrice = new Money(10);
  private _discountPolicy?: DiscountPolicy;

  withSku(sku: string): this { this._sku = sku; return this; }
  named(name: string): this { this._name = name; return this; }
  pricedAt(amount: number): this { this._unitPrice = new Money(amount); return this; }
  withDiscount(policy: DiscountPolicy): this { this._discountPolicy = policy; return this; }

  build(): Product { return new Product(this._sku, this._name, this._unitPrice, this._discountPolicy); }
}
